pkg/secretless/plugin/sharedobj/available_plugins_test.go

```diff
@@ -1,60 +1,177 @@
+		expectedTCP := mock.TCPInternalPluginsByID()
 
-var mockTCPPlugins = map[string]tcp.Plugin{
-	"one":   mockTCPPlugin{},
-	"two":   mockTCPPlugin{},
-	"three": mockTCPPlugin{},
-}
+		plugins := &Plugins{
+			HTTPPluginsByID: expectedHTTP,
+			TCPPluginsByID:  expectedTCP,
+		}
+
+		// Test HTTPPlugins()
+
+		actualHTTP := plugins.HTTPPlugins()
+
+		assert.NotNil(t, actualHTTP)
+		if actualHTTP == nil {
+			return
+		}
+
+		assert.Equal(t, expectedHTTP, actualHTTP)
+
+		// Test TCPPlugins()
+
+		actualTCP := plugins.TCPPlugins()
+
+		assert.NotNil(t, actualTCP)
+		if actualTCP == nil {
+			return
+		}
+
+		assert.Equal(t, expectedTCP, actualTCP)
+	})
 
-func getMockPlugins() plugin.AvailablePlugins {
-	return &Plugins{
-		HTTPPluginsByID: mockHTTPPlugins,
-		TCPPluginsByID:  mockTCPPlugins,
-	}
 }
 
-func TestPlugins(t *testing.T) {
-	t.Run("HTTPPlugins", func(t *testing.T) {
-		httpPlugins := getMockPlugins().HTTPPlugins()
+func TestAllAvailablePlugins(t *testing.T) {
+	t.Run("Assembles internal and external plugins", func(t *testing.T) {
+		allPlugins, err := AllAvailablePluginsWithOptions(
+			"",
+			"",
+			mock.GetInternalPlugins,
+			mock.GetExternalPlugins,
+			mock.NewLogger(),
+		)
+		assert.Nil(t, err)
+		if err != nil {
+			return
+		}
+
+		assert.EqualValues(t, mock.AllHTTPPlugins(), allPlugins.HTTPPlugins())
+		assert.EqualValues(t, mock.AllTCPPlugins(), allPlugins.TCPPlugins())
+	})
 
-		assert.NotNil(t, httpPlugins)
-		if httpPlugins == nil {
-			t.Fail()
+	t.Run("External plugins override same-named internal plugins", func(t *testing.T) {
+
+		// Setup our mocks
+
+		httpOverride := mock.NewHTTPPlugin("HTTP Override")
+		extHTTPPlugins := mock.ExternalPlugins().HTTPPlugins()
+		extHTTPPlugins["intHTTP2"] = httpOverride // Override key "intHTTP2"
+
+		tcpOverride := mock.NewTCPPlugin("TCP Override")
+		extTCPPlugins := mock.ExternalPlugins().TCPPlugins()
+		extTCPPlugins["intTCP1"] = tcpOverride // Override key "intTCP1"
+
+		allExternalPlugins := Plugins{
+			HTTPPluginsByID: extHTTPPlugins,
+			TCPPluginsByID:  extTCPPlugins,
+		}
+
+		getExternalPlugins := func(
+			pluginDir string,
+			checksumfile string,
+			logger log_api.Logger,
+		) (plugin.AvailablePlugins, error) {
+			return &allExternalPlugins, nil
+		}
+
+		// Create the test subject
+
+		allPlugins, err := AllAvailablePluginsWithOptions(
+			"",
+			"",
+			mock.GetInternalPlugins,
+			getExternalPlugins,
+			mock.NewLogger(),
+		)
+
+		// Test
+
+		assert.Nil(t, err)
+		if err != nil {
+			return
+		}
+
+		// Define our expected results
+		expectedHTTP := mock.AllHTTPPlugins()
+		expectedHTTP["intHTTP2"] = httpOverride
+		expectedTCP := mock.AllTCPPlugins()
+		expectedTCP["intTCP1"] = tcpOverride
+
+		// Ensure both the override and everything else worked
+		assert.Equal(t, expectedHTTP, allPlugins.HTTPPlugins(), "HTTP override failed")
+		assert.Equal(t, expectedTCP, allPlugins.TCPPlugins(), "TCP override failed")
```

Comment:
Just a nit but this is where a sanity check would be useful (`assert.NotEqual` before you reassign the plugins here (L121/L123))